eclipse-theia/theia
# GH-11159 Use msgpackR for default encoding of rpc messages

#### What it does

- Refactor `RpcMessageEncoder/RpcMessageDecoder`
   - Extract generic  interfaces
   - Provide default implementation based on msgpackR
   - Rename existing "old" implementation into `RecursiveMessageEncoder/Decoder`
   - Update message encoder test cases
- Introduce reusable `AbstractChannel` implementation to reduce boilerplate code for channel setup

Fixes #11159

Contributed on behalf of STMicroelectronics

#### How to test
Backend<->Frontend RPC commuication should work as before.
It's probably a good idea to test against the bugs we encountered with the "old" RpcMessageEncoder:
- #11396: Covered by `message-encoder.spec.ts` unit test
- #11249: Covered by `message-encoder.spec.ts` unit test 
- #1196: Messages should no longer appear in the console
- #11263: 
  1. Open a .spec.ts file in the repo and try the Run Mocha Tests debug configuration.
  2. Debug view should work as expected. (Variables are shown, debug, debug trace, breakpoins are hit etc.
  3. Try a dynamic Node configuration like 'Run current file'. Should also work as expected.



#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
